### ISSUE TITLE
ConsensusClient subscribe is limited to only startTime customization for a topic

### ISSUE THREAD

**Issue by Nana-EC** [hashgraph/hedera-sdk-java#339]
Currently the ConsensusClient.subscribe() at most takes in a ConsensusTopicId topicId, Instant consensusStartTime and Consumer<ConsensusMessage> listener.

This limits subscriptions to only customize the topicId and startTime of messages they want to get.
The ConsensusTopicQuery contains other properties such as EndTime and Limit that are valuable and are available in the gRPC consensus service.

This method should be updated to support these other options in the sdk.

Suggestion : change 
`subscribe(ConsensusTopicId topicId, Instant consensusStartTime, Consumer<ConsensusMessage> listener)`
to 
`subscribe(ConsensusTopicId topicId, ConsensusTopicQuery topicQuery, Consumer<ConsensusMessage> listener)`
ConsensusTopicQuery doesn't have to be the proto class, it could be some wrapper class that allows the specifications of properties.

**Comment by mehcode** [comment]
@Nana-EC Good timing. We were discussing that this morning actually and came up with the following revised API:

```java
MirrorClient mirrorClient = new MirrorClient(".....");

mirrorClient
    .forTopicId(topicId)
    .withStartTime(startTime)
    .withLimit(10)
    .subscribe((event) -> {
        // events
    });
```

I think this should satisfy what you're asking for and be extensible to later additions.

//cc @gregscullard @mike-burrage-hedera @abonander

**Comment by steven-sheehy** [comment]
@mehcode If we add a new method called `fooBar()` to mirror node API with entirely different input parameters, the solution you suggest wouldn't be extensible for that. It's only extensible for new fields in the existing method. We can add new methods that aren't even topic related in the future.

Why not just use the builder pattern to create the request object and pass it to `mirrorClient.subscribe(request, message -> {});`?

**Comment by mehcode** [comment]
> If we add a new method called fooBar() to mirror node API with entirely different input parameters, the solution you suggest wouldn't be extensible for that

Can you help me see how it wouldn't work?

Let me walk through the example again and I'll annotate my thoughts.

```java
mirrorClient
    .forTopicId(topicId) // selects [ConsensusTopicQuery]
    .withStartTime(startTime) // [ConsensusTopicQuery].[consensusStartTime]
    .withLimit(10) // [ConsensusTopicQuery].[limit]
    .subscribe((event) -> {
        // events
    });
```

---

Now with that written out I can see how this doesn't mesh as well with the existing API surface like I thought. How about this instead:

```java
var query = new ConsensusTopicQuery()
    .setTopicId(topicId)
    .setConsensusStartTime(startTime)
    .setConsensusEndTime(endTime)
    .setLimit(100);

query.subscribe(mirrorClient, message -> { ... });

// We could also support a blocking call for historical requests
query.execute(mirrorClient)
```

**Comment by steven-sheehy** [comment]
```java
mirrorClient
    .forTopicId(topicId) // selects [ConsensusTopicQuery]
...
```
This wouldn't work because it precludes us from adding another API that also takes in a topic ID. The new API you propose looks fine, though. I think the blocking API would just make things confusing though.

**Comment by mehcode** [comment]
> This wouldn't work because it precludes us from adding another API that also takes in a topic ID.

That's fair. I didn't consider that.

**Comment by mehcode** [comment]
New API is implemented in v1.1
